src/plugins/help.rs

```diff
@@ -0,0 +1,68 @@
+use crate::prelude::*;
+
+pub struct HelpPlugin {}
+
+impl HelpPlugin {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    async fn handle_help(&self, ctx: &Context, argument: Option<&str>) -> Result<()> {
+        let response = ctx.registered_commands().await?;
+
+        match argument {
+            Some(command) => match response.commands.get(command) {
+                Some(metadata) => {
+                    ctx.mention_reply(&metadata.short_help).await?;
+                }
+                None => {
+                    ctx.mention_reply(&format!("unknown command \"{}\"", command))
+                        .await?;
+                }
+            },
+            None => {
+                // Print all available commands
+                let mut commands: Vec<String> =
+                    response.commands.into_iter().map(|(k, _)| k).collect();
+                commands.sort();
+
+                ctx.mention_reply(&format!("available commands: {}", commands.join(", ")))
+                    .await?;
+                ctx.mention_reply("run \"help <command>\" for more information on a command")
+                    .await?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Plugin for HelpPlugin {
+    fn new_from_env() -> Result<Self> {
+        Ok(HelpPlugin::new())
+    }
+
+    fn command_metadata(&self) -> Vec<CommandMetadata> {
+        vec![CommandMetadata {
+            name: "help".to_string(),
+            short_help: "gives usage for a command".to_string(),
+            full_help: "gives usage and other information for a command".to_string(),
+        }]
+    }
+
+    async fn run(self, bot: Arc<Client>) -> Result<()> {
```

Comment:
random q, why do we pass in a full Arc? that mean's each run requires a ref count bump? why not just &Client from an arc Deref?

Reply:
I believe there are times when we want to start a background task from a plugin which requires an Arc. Also, this isn't as bad as you'd expect - it's one reference per plugin and should only happen when starting up or spawning  task.